src/Drupal/Component/Utility/DriverNameMatcher.php

```diff
@@ -0,0 +1,272 @@
+   * are the items labels.
+   *
+   * @var array
+   */
+    protected $candidates;
+
+  /**
+   * A string that may precede the candidate's machine names and should be
+   * ignored for identification purposes.
+   *
+   * @var string
+   */
+    protected $prefix;
+
+  /**
+   * A set of successfully matched items
+   *
+   * @var array
+   */
+    protected $results;
+
+  /**
+   * Construct a driver name matcher object.
+   *
+   * @param array $candidates
+   *   A set of candidate items where the keys are the items labels
+   *   and the values are the items machine names.
+   * @param string $prefix
+   *   A string that may precede the candidate's machine names and should be
+   *   ignored for identification purposes.
+   */
+    public function __construct($candidates, $prefix = null)
+    {
+        if (is_array($candidates)) {
+            $this->candidates = $candidates;
+        } else {
+            throw new \Exception("Candidates for identification must be passed as an array with the machine names as the keys and the labels as the values.");
+        }
+
+        $this->prefix = $prefix;
+        $this->results = [];
+    }
+
+  /**
+   * Identifies a target from the pool of candidates.
+   *
+   * @param string $target
+   *   A single string needing to be identified as an item in the candidates.
+   *
+   * @return string
+   *   The machine name of the matching candidate, or NULL if none matched.
+   */
+    public function identify($target)
+    {
+        // Wrap the target in the appropriate array for identifySet().
+        $targets = [$target => $target];
+        $results = $this->identifySet($targets);
+        // Get the first key of the results.
+        reset($results);
+        return key($results);
+    }
+
+    /**
+   * Identifies the targets from the pool of candidates.
+   *
+   * @param array $targets
+   *   A set of items needing to be identified. The key is some human-friendly
+   *   name, the value is preserved and is not used for identification.
+   *
+   * @return array
+   *   For each matched target, the key will be replaced with the machine name
+   *  of the matching candidate, & the value will be preserved. Order may vary.
+   */
+    public function identifySet($targets)
+    {
+        if (is_array($targets)) {
+            $this->targets = $targets;
+        } else {
+            throw new \Exception("Targets to be identified must be passed as an array with their human-friendly name as the keys and anything as the values.");
+        }
+
+        $mayHavePrefix = !is_null($this->prefix);
+        $this->identifyByMethod("MachineNameExactly");
+        $this->identifyByMethod("LabelExactly");
+        if ($mayHavePrefix) {
+            $this->identifyByMethod("MachineNameWithoutPrefix");
+        }
+        $this->identifyByMethod("MachineNameWithoutUnderscores");
+        if ($mayHavePrefix) {
+            $this->identifyByMethod("MachineNameWithoutPrefixAndUnderscores");
+        }
```

Comment:
The logic above here could use some code comments I think. It would be hard to grok 6-months, or a year, down the road I think.